libs/platform/ebpf_hash_table.c

```diff
@@ -214,9 +239,194 @@ _ebpf_hash_table_bucket_entry(size_t key_size, _In_ ebpf_hash_bucket_header_t* b
+    if (!local_new_bucket) {
+        result = EBPF_NO_MEMORY;
+        goto Done;
+    }
+
+    // Allocate a new backup bucket.
+    if (old_bucket_size) {
+        backup_bucket = hash_table->allocate(old_bucket_size);
+        if (!backup_bucket) {
+            result = EBPF_NO_MEMORY;
+            goto Done;
+        }
+        backup_bucket->count = old_bucket->count;
+    }
+
+    // Copy old bucket into new bucket.
+    memcpy(local_new_bucket, old_bucket, old_bucket_size);
+
+    // Append new key, data, and backup bucket.
+    ebpf_hash_bucket_entry_t* entry =
+        _ebpf_hash_table_bucket_entry(hash_table->key_size, local_new_bucket, local_new_bucket->count);
+
+    entry->backup_bucket = backup_bucket;
+    backup_bucket = NULL;
+    entry->data = data;
+    memcpy(entry->key, key, hash_table->key_size);
+    local_new_bucket->count++;
+
+    *new_bucket = local_new_bucket;
+    local_new_bucket = NULL;
+
+    result = EBPF_SUCCESS;
+
+Done:
+    hash_table->free(local_new_bucket);
+    hash_table->free(backup_bucket);
+
+    return result;
+}
+
+/**
+ * @brief Build a replacement bucket with the given entry removed.
+ * Caller must free the old bucket.
+ * Memory from the last entry's backup_bucket is used to build the new bucket.
+ * Caller must ensure that the entry is in the bucket.
+ * Operation can't fail.
+ *
+ * @param[in] hash_table The hash table.
+ * @param[in] old_bucket The immutable old bucket to copy.
+ * @param[in] key_index Location of the key to remove.
+ * @param[out] new_bucket The new bucket with the entry removed. On success the caller owns this memory.
+ */
+static void
+_ebpf_hash_table_bucket_delete(
+    _In_ ebpf_hash_table_t* hash_table,
+    _In_ ebpf_hash_bucket_header_t* old_bucket,
+    _In_ size_t key_index,
+    _Outptr_result_maybenull_ ebpf_hash_bucket_header_t** new_bucket)
+{
+    ebpf_hash_bucket_header_t* backup_bucket =
+        _ebpf_hash_table_bucket_entry(hash_table->key_size, old_bucket, old_bucket->count - 1)->backup_bucket;
+
+    // Delete the bucket if removing last entry.
+    if (old_bucket->count == 1) {
+        ebpf_assert(backup_bucket == NULL);
+        *new_bucket = backup_bucket;
+        return;
+    }
+    ebpf_assert(backup_bucket->count == old_bucket->count - 1);
+
+    // Reset bucket entry count.
+    backup_bucket->count = 0;
+
+    // Copy key and value from each entry into the backup bucket.
+    for (size_t index = 0; index < old_bucket->count; index++) {
+        if (index == key_index) {
+            continue;
+        }
+        const ebpf_hash_bucket_entry_t* old_entry =
+            _ebpf_hash_table_bucket_entry(hash_table->key_size, old_bucket, index);
+        ebpf_hash_bucket_entry_t* new_entry =
+            _ebpf_hash_table_bucket_entry(hash_table->key_size, backup_bucket, backup_bucket->count);
+
+        new_entry->data = old_entry->data;
+        memcpy(new_entry->key, old_entry->key, hash_table->key_size);
```

Comment:
If say the bucket had 10 entries (index 0-9), each entry will have a backup bucket of size n-1 where n is the index of the entry.

If we delete, for example, entry at index 5, then entries at index 6-9 will be shifted left in the new bucket, and their indices will decrease by 1. But their backup buckets still be of a larger size. 

Will that cause any issues?

Reply:
See the lines below this. Backup_buckets for entries 1 through N-1 are copied from old_bucket to new_bucket, in order.

The assert below in the code:
```
        ebpf_assert(index == 0 || new_entry->backup_bucket);
        ebpf_assert(index == 0 || new_entry->backup_bucket->count == index);
```
Verifies that each entry has an appropriately sized bucket associated with it.